### PR TITLE
Removed redundant struct bounds and unnecessary data copying

### PR DESCRIPTION
**Pull request by debris**

### CODE REVIEW

File: ethcore/src/block.rs [hunk 1]
```diff
@@ -488,11 +507,11 @@ impl<'x> IsBlock for OpenBlock<'x> {
 	fn block(&self) -> &ExecutedBlock { &self.block }
 }
 
-impl<'x> IsBlock for ClosedBlock {
```
Comment (debris) [inline ethcore/src/block.rs]: redundant lifetime

File: ethcore/src/client/client.rs [hunk 2]
```diff
@@ -345,7 +344,7 @@ impl Importer {
 		imported
 	}
 
-	fn check_and_close_block(&self, block: PreverifiedBlock, client: &Client) -> Result<LockedBlock, ()> {
+	fn check_and_lock_block(&self, block: PreverifiedBlock, client: &Client) -> Result<LockedBlock, ()> {
```
Comment (debris) [inline ethcore/src/client/client.rs]: renamed, cause the function returns `LockedBlock`, not `ClosedBlock`

File: ethcore/src/client/client.rs [hunk 3]
```diff
@@ -459,27 +458,25 @@ impl Importer {
...
-	fn commit_block<B>(&self, block: B, header: &Header, block_data: &[u8], client: &Client) -> ImportRoute where B: IsBlock + Drain {
+	fn commit_block<B>(&self, block: B, header: &Header, block_data: &[u8], client: &Client) -> ImportRoute where B: Drain {
 		let hash = &header.hash();
 		let number = header.number();
 		let parent = header.parent_hash();
 		let chain = client.chain.read();
 
 		// Commit results
-		let receipts = block.receipts().to_owned();
-		let traces = block.traces().clone().drain();
```
Comment (debris) [inline ethcore/src/client/client.rs]: receipts and traces were copied

File: util/using_queue/src/lib.rs [hunk 4]
```diff
@@ -19,7 +19,7 @@
 /// Special queue-like datastructure that includes the notion of
 /// usage to avoid items that were queued but never used from making it into
 /// the queue.
-pub struct UsingQueue<T> where T: Clone {
```
Comment (debris) [inline util/using_queue/src/lib.rs]: redundant struct bounds

File: ethcore/src/block.rs [hunk 5]
```diff
@@ -14,7 +14,22 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
-//! Blockchain block.
+//! Base data structure of this module is `Block`.
+//!
+//! Blocks can be produced by a local node or they may be received from network.
```
Comment (dvdplm) [inline ethcore/src/block.rs]: "…**the** network."

File: ethcore/src/block.rs [hunk 6]
```diff
@@ -14,7 +14,22 @@
...
+//!
+//! To create a block locally, we start with an `OpenBlock`. This block is mutable
+//! and can be appended with transactions and uncles.
```
Comment (dvdplm) [inline ethcore/src/block.rs]: "…can be appended **to** with transactions and uncles".

File: ethcore/src/block.rs [hunk 7]
```diff
@@ -14,7 +14,22 @@
...
+//!
+//! When ready, `OpenBlock` can be closed and turned into `ClosedBlock`. `ClosedBlock` can
```
Comment (dvdplm) [inline ethcore/src/block.rs]: "…turned into **a** `ClosedBlock`. **A** `ClosedBlock` can…"

File: ethcore/src/block.rs [hunk 8]
```diff
@@ -14,7 +14,22 @@
...
+//! be reopend again by miner under certain circumstances. On block close, state commit is
```
Comment (dvdplm) [inline ethcore/src/block.rs]: "…by **a** miner…"

File: ethcore/src/block.rs [hunk 9]
```diff
@@ -169,20 +194,14 @@ pub trait IsBlock {
...
-
 	/// Get all uncles in this block.
 	fn uncles(&self) -> &[Header] { &self.block().uncles }
-
-	/// Get tracing enabled flag for this block.
-	fn tracing_enabled(&self) -> bool { self.block().traces.is_enabled() }
 }
 
-/// Trait for a object that has a state database.
+/// Trait for a object that owns an `ExecutedBlock`
```
Comment (dvdplm) [inline ethcore/src/block.rs]: "…for **an** object…"

File: ethcore/src/block.rs [hunk 10]
```diff
@@ -599,8 +617,8 @@ impl SealedBlock {
 
 impl Drain for SealedBlock {
 	/// Drop this object and return the underlieing database.
```
Comment (dvdplm) [inline ethcore/src/block.rs]: "…**underlying** database"

Also: is it true that it's a "database" being returned with this change?

File: ethcore/src/client/client.rs [hunk 11]
```diff
@@ -459,27 +458,25 @@ impl Importer {
...
-		let ancestry_actions = self.engine.ancestry_actions(block.block(), &mut chain.ancestry_with_metadata_iter(*parent));
+		let ancestry_actions = self.engine.ancestry_actions(&block, &mut chain.ancestry_with_metadata_iter(*parent));
 
+		let receipts = block.receipts;
+		let traces = block.traces.drain();
 		let best_hash = chain.best_block_hash();
-		let metadata = block.block().metadata().map(Into::into);
-		let is_finalized = block.block().is_finalized();
+		let metadata = block.metadata;
+		let is_finalized = block.is_finalized;
```
Comment (dvdplm) [inline ethcore/src/client/client.rs]: Do you really need to make new bindings like this? Maybe it's more readable to see `block.is_finalized` further down the code, makes it obvious where it's coming from if one reads only parts of the method?

File: ethcore/src/block.rs (hunk 10 above)
Comment (debris) [reply ethcore/src/block.rs]: it is! good catch!

File: ethcore/src/block.rs [hunk 12]
```diff
@@ -14,7 +14,22 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
-//! Blockchain block.
+//! Base data structure of this module is `Block`.
+//!
+//! Blocks can be produced by a local node or they may be received from the network.
+//!
+//! To create a block locally, we start with an `OpenBlock`. This block is mutable
+//! and can be appended to with transactions and uncles.
```
Comment (niklasad1) [inline ethcore/src/block.rs]: can be appended ~~to~~ with transactions and uncles.

File: ethcore/src/block.rs [hunk 13]
```diff
@@ -14,7 +14,22 @@
...
+//!
+//! When ready, `OpenBlock` can be closed and turned into a `ClosedBlock`. A `ClosedBlock` can
+//! be reopend again by a miner under certain circumstances. On block close, state commit is
```
Comment (niklasad1) [inline ethcore/src/block.rs]: s/reopend/reopened

Comment (debris) [reply ethcore/src/block.rs]: I will fix this typo in #9117

**Pull request merged**